genvm/templates/multisig/multisig.go

```diff
@@ -0,0 +1,59 @@
+package multisig
+
+import (
+	"fmt"
+
+	"github.com/oasisprotocol/curve25519-voi/primitives/ed25519"
+	"github.com/spacemeshos/go-scale"
+
+	"github.com/spacemeshos/go-spacemesh/genvm/core"
+)
+
+//go:generate scalegen
+
+// MultiSig K/N template.
+type MultiSig struct {
+	k          uint8
+	PublicKeys []core.PublicKey
+}
+
+// MaxSpend returns amount specified in the SpendArguments.
+func (ms *MultiSig) MaxSpend(method uint8, args any) (uint64, error) {
+	switch method {
+	case methodSpawn:
+		return 0, nil
+	case methodSpend:
+		return args.(*SpendArguments).Amount, nil
+	default:
+		return 0, fmt.Errorf("%w: unknown method %d", core.ErrMalformed, method)
+	}
+}
+
+// Verify that transaction is signed has k valid signatures.
+func (ms *MultiSig) Verify(ctx *core.Context, raw []byte, dec *scale.Decoder) bool {
+	sig := make(Signatures, ms.k)
+	n, err := scale.DecodeStructArray(dec, sig)
+	if err != nil {
+		return false
+	}
+	hash := core.Hash(raw[:len(raw)-n])
+	batch := ed25519.NewBatchVerifierWithCapacity(int(ms.k))
+	last := uint8(0)
+	for i, part := range sig {
+		if part.Ref >= uint8(len(ms.PublicKeys)) {
+			return false
+		}
+		if i != 0 && part.Ref <= last {
```

Comment:
do you mean the signatures have to be in increasing order of Ref? why? is it a restriction on ed25519?